### DISCUSSION TITLE
The performance between c++ script and python script

### DISCUSSION THREAD

**Discussion by Junang-Wang** [sofa-framework/sofa#4984] [SofaPython3] [answered]
Hi there,

I'm currently working on a real-time simulator project and I've found the SofaPython3 plugin to be extremely helpful in reducing the learning curve for me. I'm curious to know if writing scenes in Python is significantly slower than in C++, and if the choice of complie language matters a lot. In what scenarios would it be more beneficial to choose C++?

**Comment by damienmarchal** [comment, marked as answer]
Hi @Junang-Wang 

Thank for your question and the feedback. 

In most of the cases writing a sofa scene using SofaPython3 does not have impact on performance compared to writing it with xml or c++ (with the "simpleapi.h"). This is because writing a scene consists in fact in assembling existing components and that the assembled component are in fact written in c++.  So if you are more confident with python than with c++, do with it to write scenes. 

Nevertheless, in the specific case of having a scene with *custom components* written in python (like [python controller](https://sofapython3.readthedocs.io/en/latest/content/modules/Sofa/generated/_autosummary/_autosummary/Sofa.Core.Controller.html#Sofa.Core.Controller) or [forcefield](https://github.com/sofa-framework/SofaPython3/blob/master/examples/example-forcefield.py)) ... then there can be a difference between python and c++ performance on computation intensive part of the code (if any).
 
When performances matters, the general recommendation would be to:
- do the simplest first and avoid "premature optimization"
- if performances are not enough then [trace & monitor the timing](https://sofa-framework.github.io/doc/using-sofa/performances/inspect-performances/) to spot where the bootleneck  are
- put time and efforts where there is value to that. 

And regarding Sofa, here is what I'm doing: 
- always write the scene in python (if you cannot, then use "simpleapi.h") 
- write controllers in python, if any
- write computation intensive components (forcefield, solvers, controller) in c++ after having done a sketch in python 
 
There is also this page related to how to get good simulation performances:
 https://sofa-framework.github.io/doc/using-sofa/performances/improve-performances/#profile-the-simulation.

**Reply by Junang-Wang** [reply]
Thank you for your clear response. It's very helpful.
